core/process.py

```diff
@@ -494,8 +495,12 @@ def Exec(self, arg_vec, environ):
     try:
       os_.execvpe(argv[0], argv, environ)
     except OSError as e:
-      # TODO: Run with /bin/sh when ENOEXEC error (noshebang).  Because all
-      # shells do it.
+      # Run with /bin/sh when ENOEXEC error (no shebang). Because all shells do it.
+      if e.errno == errno.ENOEXEC:
+          arg_vec.strs = (["/bin/sh"]
+            + [_ResolveFile(arg_vec.strs[0], environ["PATH"].split(':'))[1]]
```

Comment:
This looks like the right idea, but it's a bit hard to read because of the [1] and so forth.
_ResolveFile can return (None, None).  It probably won't happen, but there's no guarantee because files can disappear at any time.

It can just be argv[0] too.  It might be clearer if we make a private version of this function that's like

_Exec(self, argv, argv0_spid):
  ...

and then it can call itself in the ENOEXEC case.

Reply:
Good catch that _ResolveFile can return `None`! That found another bug - _ResolveFile wasn't handling relative paths properly, which I fixed and added to the spec tests.

I cleaned up the code a little, hopefully it's a bit easier to read now.